pulsar-client/src/main/java/org/apache/pulsar/client/impl/ConsumerImpl.java

```diff
@@ -549,13 +600,33 @@ protected CompletableFuture<Void> doAcknowledge(MessageId messageId, AckType ack
             return doTransactionAcknowledgeForResponse(messageId, ackType, null, properties,
                     new TxnID(txn.getTxnIdMostBits(), txn.getTxnIdLeastBits()));
         }
-        return acknowledgmentsGroupingTracker.addAcknowledgment((MessageIdImpl) messageId, ackType, properties);
+        if (ackType == AckType.Individual) {
+            onAcknowledge(messageId, null);
+        } else {
+            onAcknowledgeCumulative(messageId, null);
+        }
+        if (messageId instanceof BatchMessageIdImpl) {
+            BatchMessageIdImpl batchMessageId = (BatchMessageIdImpl) messageId;
+            MessageIdImpl messageIdImpl = getMessageIdToAcknowledge(batchMessageId, ackType);
```

Comment:
Would it be better to change the name of this method to AcknowledgeAndGet? Just like Atomic's IncreaseAndGet

Reply:
No. This method is to get the `MessageIdImpl` to acknowledge. `increaseAndGet` modifies the atomic variable itself, but `getMessageIdToAcknowledge` just calculated a `MessageIdImpl` to acknowledge from the `BatchMessageIdImpl` and does not modify the `BatchMessageIdImpl` itself.